File: python/tests/test_tier4_batched.py
"""Tests for v3 Tier-4 (Monte-Carlo) batched dispatch:
``balance`` and ``evenness``.

The new wrinkle vs Tiers 1/2 is RNG handling. The ``rng_scope``
argument controls how each row's seed is derived from the base
``rng_seed``:

- ``'canonical'`` (default): per-row seed = ``base + fnv1a32(canonical_key)``.
  Canonical-form-equivalent rows get identical seeds and identical
  Monte-Carlo realisations, so dedup works for ``sigma > 0``.
- ``'row'``: per-row seed = ``base + row_index``. Each row gets an
  independent reproducible realisation; dedup is disabled.

When ``rng_seed`` is None in batched mode, a session-random base is
generated once per call so within-call dedup is reproducible while
across-call results differ.
"""

import numpy as np
import pytest

from mpt import balance, evenness


# ---------------------------------------------------------------------
# sigma = 0 (deterministic)
# ---------------------------------------------------------------------


class TestBalanceSigmaZero:
    def test_scalar_unchanged(self):
        b = balance([0, 4, 7], None, 12)
        assert isinstance(b, float)

    def test_batched_returns_array(self):
        P = np.array([[0, 4, 7], [0, 3, 7]])
        b = balance(P, None, 12)
        assert b.shape == (2,)

    def test_batched_matches_scalar(self):
        P = np.array([[0, 4, 7], [0, 3, 7]])
        b = balance(P, None, 12)
        assert abs(b[0] - balance([0, 4, 7], None, 12)) < 1e-12

    def test_batched_dedup_permutation(self):
        P = np.array([[0, 4, 7], [4, 0, 7], [7, 4, 0]])
        b = balance(P, None, 12)
        assert b[0] == b[1] == b[2]

    def test_batched_with_return_std(self):
        P = np.array([[0, 4, 7], [0, 3, 7]])
        b, bs = balance(P, None, 12, return_std=True)
        assert b.shape == bs.shape == (2,)
        assert np.all(bs == 0.0)  # sigma=0 -> std is 0

    def test_batched_nan_padded(self):
        P = np.array([[0, 4, 7, np.nan], [0, 1, 5, 6], [np.nan]*4])
        b = balance(P, None, 12)
        assert not np.isnan(b[0]) and not np.isnan(b[1]) and np.isnan(b[2])


class TestEvennessSigmaZero:
    def test_batched_returns_array(self):
        P = np.array([[0, 4, 7], [0, 3, 7]])
        e = evenness(P, 12)
        assert e.shape == (2,)

    def test_batched_dedup(self):
        P = np.array([[0, 4, 7], [4, 0, 7]])
        e = evenness(P, 12)
        assert e[0] == e[1]


# ---------------------------------------------------------------------
# sigma > 0 (Monte Carlo) — rng_scope contract
# ---------------------------------------------------------------------


class TestBalanceMonteCarloCanonicalScope:
    """Default rng_scope='canonical': identical canonical keys give
    identical MC realisations within a single call."""

    def test_identical_inputs_identical_results(self):
        P = np.array([[0, 4, 7], [4, 0, 7], [0, 4, 7]])
        b = balance(P, None, 12, sigma=0.3, rng_seed=42)
        assert b[0] == b[1] == b[2]

    def test_different_canonicals_different_results(self):
        # Major and minor triads are distinct multisets — different
        # canonical keys -> different derived seeds -> different MC.
        P = np.array([[0, 4, 7], [0, 3, 7]])
        b = balance(P, None, 12, sigma=0.3, rng_seed=42)
        assert b[0] != b[1]

    def test_reproducible_across_calls(self):
        P = np.array([[0, 4, 7], [0, 3, 7]])
        b1 = balance(P, None, 12, sigma=0.3, rng_seed=42)
        b2 = balance(P, None, 12, sigma=0.3, rng_seed=42)
        assert np.array_equal(b1, b2)

    def test_different_seeds_different_results(self):
        P = np.array([[0, 4, 7], [0, 3, 7]])
        b1 = balance(P, None, 12, sigma=0.3, rng_seed=42)
        b2 = balance(P, None, 12, sigma=0.3, rng_seed=123)
        assert not np.array_equal(b1, b2)

    def test_rng_seed_none_within_call_dedup(self):
        # Within one call, dedup must work even without explicit seed.
        P = np.array([[0, 4, 7], [4, 7, 0], [0, 4, 7]])
        b = balance(P, None, 12, sigma=0.3)  # rng_seed=None
        assert b[0] == b[1] == b[2]

    def test_rng_seed_none_across_calls_differs(self):
        # Across calls without explicit seed, results should differ
        # (session-random base).
        P = np.array([[0, 4, 7]])
        b1 = balance(P, None, 12, sigma=0.3)
        b2 = balance(P, None, 12, sigma=0.3)
        # Vanishingly small chance of equality by coincidence.
        assert b1[0] != b2[0]


class TestBalanceMonteCarloRowScope:
    """rng_scope='row': identical inputs get DIFFERENT realisations."""

    def test_identical_inputs_different_results(self):
        P = np.array([[0, 4, 7], [0, 4, 7], [0, 4, 7]])
        b = balance(P, None, 12, sigma=0.3, rng_seed=42, rng_scope="row")
        # All three rows are the same scale but get distinct seeds.
        assert b[0] != b[1] and b[1] != b[2] and b[0] != b[2]

    def test_reproducible_across_calls(self):
        P = np.array([[0, 4, 7], [0, 4, 7]])
        b1 = balance(P, None, 12, sigma=0.3, rng_seed=42, rng_scope="row")
        b2 = balance(P, None, 12, sigma=0.3, rng_seed=42, rng_scope="row")
        assert np.array_equal(b1, b2)


class TestEvennessMonteCarloCanonical:
    """Spot-check evenness uses the same RNG infrastructure."""

    def test_dedup_canonical(self):
        P = np.array([[0, 4, 7], [4, 0, 7]])
        e = evenness(P, 12, sigma=0.3, rng_seed=42)
        assert e[0] == e[1]

    def test_row_independent(self):
        P = np.array([[0, 4, 7], [0, 4, 7]])
        e = evenness(P, 12, sigma=0.3, rng_seed=42, rng_scope="row")
        assert e[0] != e[1]


class TestRngScopeValidation:
    def test_invalid_rng_scope_raises(self):
        P = np.array([[0, 4, 7], [0, 3, 7]])
        with pytest.raises(ValueError, match="rng_scope"):
            balance(P, None, 12, rng_scope="invalid")
        with pytest.raises(ValueError, match="rng_scope"):
            evenness(P, 12, rng_scope="invalid")

File: matlab/tests/test_tier4_batched.m
%% test_tier4_batched.m — Tier-4 batched dispatch (v3+): balanceCircular, evennessCircular
%
%  Tests for Tier-4 batched dispatch (v3+): balanceCircular, evennessCircular.
%
%  Standalone-runnable; appends to `results` when called from
%  test_mpt.m.

if ~exist('results', 'var')
    results = {};
    standalone = true;
    addpath(fileparts(mfilename('fullpath')));
    clear cleanupDefaults
    cleanupDefaults = mptTestIsolateDefaults(); %#ok<NASGU>
else
    standalone = false;
end


%
% These Monte-Carlo functions accept an nRows-by-K matrix and return
% per-row column vectors. The new ``rngScope`` NV pair controls how
% each row's RNG seed is derived from the base ``rngSeed``:
%   'canonical' (default): derived from canonical-form key, so
%      transposition-equivalent rows share an MC realisation; dedup
%      works for sigma > 0.
%   'row': derived from row index, so identical rows get distinct
%      reproducible realisations; dedup is disabled.

% --- balanceCircular: sigma = 0 (deterministic) ---
P_b0 = [0, 4, 7; 0, 3, 7; 0, 4, 7];
bVec0 = balanceCircular(P_b0, [], 12, 0);
results{end+1,1} = 'balanceCircular batched sigma=0: returns nRows-by-1 vector';
results{end,2}   = isnumeric(bVec0) && isequal(size(bVec0), [3, 1]);

results{end+1,1} = 'balanceCircular batched sigma=0: matches scalar dispatch';
results{end,2}   = abs(bVec0(2) - balanceCircular(P_b0(2, :)', [], 12, 0)) < 1e-12;

results{end+1,1} = 'balanceCircular batched sigma=0: dedup (rows 1, 3 identical)';
results{end,2}   = bVec0(1) == bVec0(3);

[bVec0_, bStd0_] = balanceCircular(P_b0, [], 12, 0);
results{end+1,1} = 'balanceCircular batched sigma=0: bStd is zero with two outputs';
results{end,2}   = isequal(size(bStd0_), [3, 1]) && all(bStd0_ == 0);

% NaN-padded
P_b0_nan = [0, 4, 7, NaN; 0, 1, 5, 6; NaN, NaN, NaN, NaN];
bNan = balanceCircular(P_b0_nan, [], 12, 0);
results{end+1,1} = 'balanceCircular batched sigma=0: NaN-padded rows give NaN';
results{end,2}   = ~isnan(bNan(1)) && ~isnan(bNan(2)) && isnan(bNan(3));

% --- balanceCircular: sigma > 0 (Monte Carlo) ---

% Canonical scope: identical inputs give identical MC results
P_b1 = [0, 4, 7; 4, 0, 7; 0, 4, 7];
bCanon = balanceCircular(P_b1, [], 12, 0.3, ...
    'rngSeed', 42, 'rngScope', 'canonical');
results{end+1,1} = 'balanceCircular batched MC canonical: identical inputs identical results';
results{end,2}   = bCanon(1) == bCanon(2) && bCanon(1) == bCanon(3);

% Different canonicals -> different results
P_b2 = [0, 4, 7; 0, 3, 7];
bDiff = balanceCircular(P_b2, [], 12, 0.3, ...
    'rngSeed', 42, 'rngScope', 'canonical');
results{end+1,1} = 'balanceCircular batched MC canonical: different scales give different results';
results{end,2}   = bDiff(1) ~= bDiff(2);

% Reproducibility across calls
b_call_a = balanceCircular(P_b1, [], 12, 0.3, 'rngSeed', 42);
b_call_b = balanceCircular(P_b1, [], 12, 0.3, 'rngSeed', 42);
results{end+1,1} = 'balanceCircular batched MC: reproducible with same rngSeed';
results{end,2}   = isequal(b_call_a, b_call_b);

% Row scope: identical inputs give DIFFERENT realisations
bRow = balanceCircular(P_b1, [], 12, 0.3, ...
    'rngSeed', 42, 'rngScope', 'row');
results{end+1,1} = 'balanceCircular batched MC row scope: identical inputs differ';
results{end,2}   = bRow(1) ~= bRow(2) && bRow(2) ~= bRow(3) && bRow(1) ~= bRow(3);

% Empty rngSeed in batched: within-call dedup still works
bDedup = balanceCircular(P_b1, [], 12, 0.3);
results{end+1,1} = 'balanceCircular batched MC: within-call dedup with empty rngSeed';
results{end,2}   = bDedup(1) == bDedup(2) && bDedup(1) == bDedup(3);

% --- evennessCircular: brief MC sanity ---
P_e = [0, 4, 7; 4, 0, 7];
eDedup = evennessCircular(P_e, 12, 0.3, 'rngSeed', 42, 'rngScope', 'canonical');
results{end+1,1} = 'evennessCircular batched MC canonical: dedup works';
results{end,2}   = eDedup(1) == eDedup(2);

eRow = evennessCircular([0, 4, 7; 0, 4, 7], 12, 0.3, ...
    'rngSeed', 42, 'rngScope', 'row');
results{end+1,1} = 'evennessCircular batched MC row scope: identical inputs differ';
results{end,2}   = eRow(1) ~= eRow(2);


%% ---- Standalone summary ----

if standalone
    nPass = sum([results{:, 2}]);
    nFail = size(results, 1) - nPass;
    for i = 1:size(results, 1)
        if results{i, 2}
            fprintf('  PASS  %s\n', results{i, 1});
        else
            fprintf('  FAIL  %s\n', results{i, 1});
        end
    end
    fprintf('\n=== test_tier4_batched: %d passed, %d failed (of %d) ===\n\n', ...
        nPass, nFail, nPass + nFail);
    clear cleanupDefaults
    if nFail > 0
        error('test_tier4_batched:failed', '%d test(s) failed.', nFail);
    end
end
